File: golang/version1/main.go
package main

import (
	"fmt"
	"math/rand"
	"sort"
	"time"
)

func main() {
	fmt.Println("Generating a new deck.")
	for x := 0; x < 100; x++ {
		card_deck := new_deck()
		for i := 0; i < 26; i++ {
			shuffle_deck(&card_deck)
			deal_cards(&card_deck)
		}
	}

}

func new_deck() []string {
	my_deck := make([]string, 0)
	for _, s := range []string{"S", "D", "C", "H"} {
		for _, v := range []string{"A", "1", "2", "3", "4", "5", "6", "7", "8", "9", "J", "Q", "K"} {
			my_deck = append(my_deck, fmt.Sprintf("%s%s", v, s))
		}
	}
	return my_deck
}

func shuffle_deck(card_deck *[]string) {
	rand.Seed(time.Now().UTC().UnixMicro())
	sort.Slice(*card_deck, func(a, b int) bool { return rand.Intn(2) == 0 })
}

func deal_cards(my_deck *[]string) {
	choice_one := (*my_deck)[0]
	choice_two := (*my_deck)[1]
	*my_deck = (*my_deck)[2:]
	fmt.Printf("%s, %s, %d\n", choice_one, choice_two, len(*my_deck))
}

File: rust/version1/src/main.rs
use rand::thread_rng;
use rand::seq::SliceRandom;

const SUITS: [char; 4] = ['S','D','C','H'];
const VALUES: [char; 13] = ['A','1','2','3','4','5','6','7','8','9','J','Q','K'];
fn new_deck() -> Vec<String> {
    let mut my_deck: Vec<String> = Vec::new();
    for s in SUITS.iter() {
        for v in VALUES.iter() {
            my_deck.push(v.to_string() + &s.to_string());
        }
    }
    return my_deck;
}

fn shuffle_deck(card_deck: &mut Vec<String>) {
    card_deck.shuffle(&mut thread_rng());
}

fn deal_cards(my_deck: &mut Vec<String>) {
    let choice_one: String = my_deck.pop().unwrap();
    let choice_two: String = my_deck.pop().unwrap();
    println!("{}, {}, {}", choice_one, choice_two, my_deck.len());
}

fn main() {
    println!("Generating a new deck.");
    for _x in 0..100 {
        let mut card_deck: Vec<String> = new_deck();
        for _i in 0..26 {
            shuffle_deck(&mut card_deck);
            deal_cards(&mut card_deck);
        }
    }
}
